File: backend_nodejs_nest_fastify/my-pizza/src/nest_modules/product/controllers/product-controller.ts
import { Body, Controller, DefaultValuePipe, Delete, Get, Param, ParseIntPipe, Post, Put, Query, UseGuards } from '@nestjs/common';
import { AdminGuard } from '../../auth-user/guards/admin.guard';
import { ProductService } from '../services/product.service';
import { MenuService } from '../services/menu.service';
import { ToppingType } from '../enums/topping-type.enum';
import { ToppingsRes } from '../interfaces/toppings-res.interface';
import { ToppingDTO } from '../interfaces/topping-dto.interface';
import { ToppingRes } from '../interfaces/topping-res.interface';
import { ConfirmRes } from 'src/nest_modules/auth-user/interfaces/confirm-res.interface';
import { Pagination } from 'nestjs-typeorm-paginate';
import { ProductNamesRes } from '../interfaces/products-names-res.interface';
import { CategoriesRes } from '../interfaces/categories-res.interface';
import { ManyproductsPostDTO } from '../interfaces/many-products-post-dto.interface';
import { ProductRes } from '../interfaces/product-res.interface';
import { ProductDTO } from '../interfaces/product-dto.interface';
import { MenuDTO } from '../entities/menu-dto.interface';

@Controller('api')
@UseGuards(AdminGuard)
export class ProductController {

    constructor(private productSvc: ProductService, private menuSvc: MenuService) { }

    // -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
    // Toppings

    @Get('toppings')
    public async getAllToppings(@Query('type') type: ToppingType): Promise<ToppingsRes> {
        const isTypeDefined: boolean = !!type
        if (!isTypeDefined) {
            return {
                toppings: await this.productSvc.findAllToppings()
            }
        } else {
            return {
                toppings: await this.productSvc.findAllToppings(type)
            }
        }
    }

    @Post('toppings')
    public async addTopping(@Body() toppingDTO: ToppingDTO): Promise<ToppingRes> {
        return await this.productSvc.addTopping(toppingDTO)
    }

    @Put('toppings/:name')
    public async updateToppingByName(@Param('name') name: string, @Body() toppingDTO: ToppingDTO): Promise<ToppingRes> {
        return await this.productSvc.updateToppingByName(name, toppingDTO)
    }

    @Delete('toppings/:name')
    public async deleteToppingByName(@Param('name') name: string): Promise<ConfirmRes> {
        return await this.productSvc.deleteToppingByName(name)
    }

    // -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
    // Products

    @Get('products')
    public async getAllProducts(@Query('page', new DefaultValuePipe(1), ParseIntPipe) page: number = 1,
        @Query('limit', new DefaultValuePipe(10), ParseIntPipe) limit: number = 10): Promise<Pagination<ProductRes>> {

        return await this.productSvc.getAllProducts({ page, limit })

    }

    @Get('products/product-names')
    public async getProductNames(): Promise<ProductNamesRes> {
        return {
            productNames: await this.productSvc.getProductNames()
        }
    }

    @Get('products/categories')
    public async getCategories(): Promise<CategoriesRes> {
        return {
            categories: await this.productSvc.getAllCategoryNames()
        }
    }

    @Post('products/add-many')
    public async addManyProducts(@Body() manyProductsPostDTO: ManyproductsPostDTO): Promise<ProductRes[]> {
        return await this.productSvc.addManyProducts(manyProductsPostDTO)
    }

    @Put('products/:name')
    public async updateProduct(@Param('name') name: string, productDTO: ProductDTO): Promise<ProductRes> {
        return await this.productSvc.updateProductByName(name, productDTO)
    }

    @Delete('products/:name')
    public async deleteProduct(@Param('name') name: string): Promise<ConfirmRes> {
        return this.productSvc.deleteProductByName(name)
    }

    @Post()
    public async setMenu(menuDTO: MenuDTO): Promise<ConfirmRes> {
        return this.menuSvc.saveMenu(menuDTO.menuIds)
    }

}

File: backend_java_spring_boot/myPizza/src/main/java/backendapp/myPizza/controllers/ProductController.java
package backendapp.myPizza.controllers;

import backendapp.myPizza.Models.entities.Menu;
import backendapp.myPizza.Models.entities.Product;
import backendapp.myPizza.Models.entities.Topping;
import backendapp.myPizza.Models.enums.ToppingType;
import backendapp.myPizza.Models.reqDTO.*;
import backendapp.myPizza.Models.resDTO.CategoriesRes;
import backendapp.myPizza.Models.resDTO.ConfirmRes;
import backendapp.myPizza.Models.resDTO.ProductNamesRes;
import backendapp.myPizza.Models.resDTO.ToppingsRes;
import backendapp.myPizza.exceptions.BadRequestException;
import backendapp.myPizza.exceptions.NotFoundException;
import backendapp.myPizza.exceptions.Validation;
import backendapp.myPizza.services.MenuService;
import backendapp.myPizza.services.ProductService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.validation.BindingResult;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api")
@PreAuthorize("hasAuthority('ADMIN')")
public class ProductController {

    @Autowired
    private ProductService productSvc;

    @Autowired
    private MenuService menuSvc;

    // -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
    // Toppings

    @GetMapping("/toppings")
    public ToppingsRes getAllToppings(@RequestParam(required = false) ToppingType type) {
        switch (type) {
            case ToppingType.EXTRA -> {
                return new ToppingsRes(productSvc.findAllExtras());
            }
            case ToppingType.TOPPING -> {
                return new ToppingsRes(productSvc.findAllToppingToppings());
            }
            case null -> {
                return new ToppingsRes(productSvc.findAllToppings());
            }
        }

    }

    @PostMapping("/toppings")
    public Topping addTopping(@RequestBody @Validated ToppingDTO toppingDTO, BindingResult validation) throws BadRequestException {
        Validation.validate(validation);
        return productSvc.addTopping(toppingDTO);
    }

    @PutMapping("/toppings/{name}")
    public Topping updateToppingByName(@PathVariable String name, @RequestBody @Validated ToppingDTO toppingDTO, BindingResult validation) throws BadRequestException {
        Validation.validate(validation);
        return productSvc.updateToppingByName(name, toppingDTO);
    }

    @DeleteMapping("/toppings/{name}")
    public ConfirmRes deleteToppingByName(@PathVariable String name) throws BadRequestException {
        return productSvc.deleteToppingByName(name);
    }

    // -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
    // Products

    @GetMapping("/products")
    public Page<Product> getAllProducts(Pageable pageable) {
        return productSvc.getAllProducts(pageable);
    }

    @GetMapping("/products/product-names")
    public ProductNamesRes getProductNames() {
        return new ProductNamesRes(productSvc.getProductNames());
    }

    @GetMapping("/products/categories")
    public CategoriesRes getCategories() {
        return new CategoriesRes(productSvc.getAllCategoryNames());
    }

    @PostMapping("/products")
    public Product addProduct(@RequestBody @Validated ProductDTO productDTO, BindingResult validation) throws BadRequestException {
        Validation.validate(validation);
        return productSvc.addProduct(productDTO);
    }

    @PostMapping("/products/add-many")
    public List<Product> addManyProducts(@RequestBody @Validated ManyProductsPostDTO manyProductsPostDTO, BindingResult validation) throws BadRequestException {
        Validation.validate(validation);
        return productSvc.addManyProducts(manyProductsPostDTO);
    }

    @PutMapping("/products/{name}")
    public Product updateProduct(@PathVariable String name, @RequestBody @Validated ProductDTO productDTO, BindingResult validation) throws BadRequestException, NotFoundException {
        Validation.validate(validation);
        return productSvc.updateProductByName(name, productDTO);
    }

    @DeleteMapping("/products/{name}")
    public ConfirmRes deleteProduct(@PathVariable String name) throws BadRequestException {
        return productSvc.deleteProductByName(name);
    }

    @PostMapping("/set-menu")
    public ConfirmRes setMenu(@RequestBody @Validated MenuDTO menuDTO, BindingResult validation) throws BadRequestException {
        Validation.validate(validation);
        return menuSvc.saveMenu(menuDTO.menuIds());
    }   


}